pkg/streamer/reader_test.go

```diff
@@ -0,0 +1,702 @@
+	for {
+		ev, err2 := s.GetEvent(ctx)
+		c.Assert(err2, IsNil)
+		if ev.Header.Timestamp == 0 || ev.Header.LogPos == 0 || ev.Header.EventType == replication.FORMAT_DESCRIPTION_EVENT {
+			continue // ignore fake event and FormatDescriptionEvent, go-mysql may send extra FormatDescriptionEvent
+		}
+		obtainExtraEvents = append(obtainExtraEvents, ev)
+		if len(obtainExtraEvents) == cap(obtainExtraEvents) {
+			break
+		}
+	}
+	t.verifyNoEventsInStreamer(c, s)
+
+	// verify obtain extra events
+	c.Assert(obtainExtraEvents, DeepEquals, extraEvents)
+
+	// 3. create new file in the last directory
+	lastFilename = filepath.Join(baseDir, UUIDs[2], filenamePrefix+strconv.Itoa(4))
+	err = ioutil.WriteFile(lastFilename, eventsBuf.Bytes(), 0600)
+	c.Assert(err, IsNil)
+
+	obtainExtraEvents2 := make([]*replication.BinlogEvent, 0, len(baseEvents)-1)
+	for {
+		ev, err2 := s.GetEvent(ctx)
+		c.Assert(err2, IsNil)
+		if ev.Header.Timestamp == 0 || ev.Header.LogPos == 0 || ev.Header.EventType == replication.FORMAT_DESCRIPTION_EVENT {
+			continue // ignore fake event and FormatDescriptionEvent, go-mysql may send extra FormatDescriptionEvent
+		}
+		obtainExtraEvents2 = append(obtainExtraEvents2, ev)
+		if len(obtainExtraEvents2) == cap(obtainExtraEvents2) {
+			break
+		}
+	}
+	t.verifyNoEventsInStreamer(c, s)
+
+	// verify obtain extra events
+	c.Assert(obtainExtraEvents2, DeepEquals, baseEvents[1:])
+
+	// NOTE: load new UUIDs dynamically not supported yet
+
+	// close the reader
+	err = r.Close()
+	c.Assert(err, IsNil)
+}
+
+func (t *testReaderSuite) TestStartSyncError(c *C) {
+	var (
+		baseDir = c.MkDir()
+		UUIDs   = []string{
+			"b60868af-5a6f-11e9-9ea3-0242ac160006.000001",
+		}
+		cfg      = &BinlogReaderConfig{RelayDir: baseDir}
+		startPos = gmysql.Position{Name: "test-mysql-bin|000001.000001"} // from the first relay log file in the first sub directory
+	)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	r := NewBinlogReader(cfg)
+
+	// no startup pos specified
+	s, err := r.StartSync(gmysql.Position{})
+	c.Assert(errors.Cause(err), Equals, ErrBinlogFileNotSpecified)
+	c.Assert(s, IsNil)
+
+	// empty UUIDs
+	s, err = r.StartSync(startPos)
+	c.Assert(err, IsNil)
+	ev, err := s.GetEvent(ctx)
+	c.Assert(err, ErrorMatches, ".*empty UUIDs not valid.*")
+	c.Assert(ev, IsNil)
+	c.Assert(r.Close(), IsNil)
+
+	// write UUIDs into index file
+	r = NewBinlogReader(cfg) // create a new reader
+	uuidBytes := t.uuidListToBytes(c, UUIDs)
+	err = ioutil.WriteFile(r.indexPath, uuidBytes, 0600)
+	c.Assert(err, IsNil)
+
+	// the startup relay log file not found
+	s, err = r.StartSync(startPos)
+	c.Assert(err, IsNil)
+	ev, err = s.GetEvent(ctx)
+	c.Assert(err, ErrorMatches, fmt.Sprintf(".*%s.*not found.*", startPos.Name))
+
+	// can not re-start the reader
+	s, err = r.StartSync(startPos)
+	c.Assert(errors.Cause(err), Equals, ErrReaderRunning)
+	c.Assert(s, IsNil)
+
+	c.Assert(ev, IsNil)
```

Comment:
how about move this line after L626?

Reply:
👍 done in https://github.com/pingcap/dm/pull/140/commits/d7f60d5dcf1f9caa3fa89c0ad8bbb6821281bd1e.